pkg/virt-handler/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,123 @@
+package heartbeat
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	k8scli "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/log"
+	virtutil "kubevirt.io/kubevirt/pkg/util"
+	clusterutils "kubevirt.io/kubevirt/pkg/util/cluster"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+type HeartBeat struct {
+	clientset               k8scli.CoreV1Interface
+	deviceManagerController device_manager.DeviceControllerInterface
+	clusterConfig           *virtconfig.ClusterConfig
+	host                    string
+	cpuManagerPath          string
+}
+
+func NewHeartBeat(clientset k8scli.CoreV1Interface, deviceManager device_manager.DeviceControllerInterface, clusterConfig *virtconfig.ClusterConfig, host string, clusterType clusterutils.ClusterType) *HeartBeat {
+	// This is a temporary workaround until k8s bug #66525 is resolved
+	var cpuManagerPath string
+	switch clusterType {
+	case clusterutils.ClusterTypeOCP3:
+		cpuManagerPath = virtutil.CPUManagerOS3Path
+	default:
+		cpuManagerPath = virtutil.CPUManagerPath
+	}
+	return &HeartBeat{
+		clientset:               clientset,
+		deviceManagerController: deviceManager,
+		clusterConfig:           clusterConfig,
+		host:                    host,
+		cpuManagerPath:          cpuManagerPath,
+	}
+}
+
+func (h *HeartBeat) Run(interval time.Duration, stopCh chan struct{}) {
+	go h.heartBeat(interval, stopCh)
+}
+
+func (h *HeartBeat) heartBeat(interval time.Duration, stopCh chan struct{}) {
+	// TODO Improve this
+	// Give the device plugins some time to initialize before going into the heartbeat loop
+	// The device plugins are normally almost immediately ready, still if we do the first heartbeat just
+	// a tiny bit before the device plugins are initialized, the node is not schedulable for a whole minute.
+	_ = wait.Poll(500*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+		return h.deviceManagerController.Initialized(), nil
+	})
+	for {
+		wait.JitterUntil(h.do, interval, 1.2, true, stopCh)
+	}
+}
+
+func (h *HeartBeat) do() {
+	now, err := json.Marshal(metav1.Now())
+	if err != nil {
+		log.DefaultLogger().Reason(err).Errorf("Can't determine date")
+		return
+	}
+
+	kubevirtSchedulable := "true"
+	if !h.deviceManagerController.Initialized() {
+		kubevirtSchedulable = "false"
+	}
+
+	var data []byte
+	if h.clusterConfig.CPUManagerEnabled() {
+		// Label the node if cpu manager is running on it
+		// This is a temporary workaround until k8s bug #66525 is resolved
```

Comment:
Not related to this PR, but someone need to make sure #66525 is opened and discussed, as k8s keep closing it.